### PR TITLE
planner: remove constant sort items after substitution

### PR DESCRIPTION
**Pull request by zz-jason**
### What problem does this PR solve? 

fix https://github.com/pingcap/tidb/issues/9331

### What is changed and how it works?

remove remove meaningless constant sort items after column substitution.

### Check List 

Tests 

 - Integration test

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

**Comment by zz-jason** [comment]
This PR fixes the broken integration tests, please review it in a high priority.

File: planner/core/rule_topn_push_down.go [hunk 1]
```diff
@@ -96,6 +96,16 @@ func (p *LogicalProjection) pushDownTopN(topN *LogicalTopN) LogicalPlan {
 		for _, by := range topN.ByItems {
 			by.Expr = expression.ColumnSubstitute(by.Expr, p.schema, p.Exprs)
 		}
+
+		// remove meaningless constant sort items.
+		for i := len(topN.ByItems) - 1; i >= 0; i-- {
+			_, isConst := topN.ByItems[i].Expr.(*expression.Constant)
+			if isConst {
+				numItems := len(topN.ByItems)
+				copy(topN.ByItems[i:numItems-1], topN.ByItems[i+1:])
```
Comment (XuHuaiyu) [inline planner/core/rule_topn_push_down.go]: A more common way seems to be:
``` go
topN.ByItems = append(topN.ByItems[:i], topN.ByItems[i+1:])
```

Comment (zz-jason) [reply planner/core/rule_topn_push_down.go]: done

File: planner/core/rule_topn_push_down.go [hunk 2]
```diff
@@ -96,6 +96,14 @@ func (p *LogicalProjection) pushDownTopN(topN *LogicalTopN) LogicalPlan {
 		for _, by := range topN.ByItems {
 			by.Expr = expression.ColumnSubstitute(by.Expr, p.schema, p.Exprs)
 		}
+
+		// remove meaningless constant sort items.
+		for i := len(topN.ByItems) - 1; i >= 0; i-- {
+			_, isConst := topN.ByItems[i].Expr.(*expression.Constant)
```
Comment (eurekaka) [inline planner/core/rule_topn_push_down.go]: What if the query is like `select a, b from t order by 2`?

Comment (zz-jason) [reply planner/core/rule_topn_push_down.go]: `order by 2` is treated as `order by b`, so the execution plan of this statement is:
```sql
TiDB(root@127.0.0.1:test) > desc select a, b from t order by 2 limit 1;
+------------------------+----------+------+------------------------------------------------------------+
| id                     | count    | task | operator info                                              |
+------------------------+----------+------+------------------------------------------------------------+
| TopN_7                 | 1.00     | root | test.t.b:asc, offset:0, count:1                            |
| └─TableReader_15       | 1.00     | root | data:TopN_14                                               |
|   └─TopN_14            | 1.00     | cop  | test.t.b:asc, offset:0, count:1                            |
|     └─TableScan_13     | 10000.00 | cop  | table:t, range:[-inf,+inf], keep order:false, stats:pseudo |
+------------------------+----------+------+------------------------------------------------------------+
4 rows in set (0.00 sec)
```

**Pull request merged**